tutorials/early_stopping/early_stopping.ipynb

```diff
@@ -0,0 +1,978 @@
+    "from ax.modelbridge.dispatch_utils import choose_generation_strategy\n",
+    "from ax.runners.torchx import TorchXRunner\n",
+    "from ax.service.scheduler import Scheduler, SchedulerOptions\n",
+    "from ax.service.utils.report_utils import exp_to_df\n",
+    "from matplotlib import pyplot as plt\n",
+    "from torchx import specs\n",
+    "from torchx.components import utils\n",
+    "\n",
+    "%matplotlib inline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fe2cf6fe",
+   "metadata": {},
+   "source": [
+    "## Defining the TorchX App\n",
+    "\n",
+    "Our goal is to optimize the PyTorch Lightning training job defined in\n",
+    "[mnist_train_nas.py](https://github.com/pytorch/tutorials/tree/master/intermediate_source/mnist_train_nas.py)_.\n",
+    "To do this using TorchX, we write a helper function that takes in\n",
+    "the values of the architcture and hyperparameters of the training\n",
+    "job and creates a [TorchX AppDef](https://pytorch.org/torchx/latest/basics.html)_\n",
+    "with the appropriate settings.\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "b423923c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def trainer(\n",
+    "    log_path: str,\n",
+    "    hidden_size_1: int,\n",
+    "    hidden_size_2: int,\n",
+    "    learning_rate: float,\n",
+    "    dropout: float,\n",
+    "    trial_idx: int = -1,\n",
+    ") -> specs.AppDef:\n",
+    "\n",
+    "    # define the log path so we can pass it to the TorchX AppDef\n",
+    "    if trial_idx >= 0:\n",
+    "        log_path = Path(log_path).joinpath(str(trial_idx)).absolute().as_posix()\n",
+    "    \n",
+    "    epochs = 10\n",
+    "    batch_size = 32\n",
+    "    \n",
+    "    return utils.python(\n",
+    "        # command line args to the training script\n",
+    "        \"--log_path\",\n",
+    "        log_path,\n",
+    "        \"--hidden_size_1\",\n",
+    "        str(hidden_size_1),\n",
+    "        \"--hidden_size_2\",\n",
+    "        str(hidden_size_2),\n",
+    "        \"--learning_rate\",\n",
+    "        str(learning_rate),\n",
+    "        \"--epochs\",\n",
+    "        str(epochs),\n",
+    "        \"--dropout\",\n",
+    "        str(dropout),\n",
+    "        \"--batch_size\",\n",
+    "        str(batch_size),\n",
+    "        # other config options\n",
+    "        name=\"trainer\",\n",
+    "        script=\"mnist_train_nas.py\",\n",
+    "        image=torchx.version.TORCHX_IMAGE,\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "65f7011d",
+   "metadata": {},
+   "source": [
+    "## Setting up the Runner\n",
+    "\n",
+    "Ax’s [Runner](https://ax.dev/api/core.html#ax.core.runner.Runner)_\n",
+    "abstraction allows writing interfaces to various backends.\n",
+    "Ax already comes with Runner for TorchX, and so we just need to\n",
+    "configure it. For the purpose of this tutorial we run jobs locally\n",
```

Comment:
May make sense here to give a heads up that the training job will be logging partial results to tensorboard, which we will read while the training is ongoing in order to implement trial-level early stopping.